library/src/main/java/com/datatorrent/lib/appdata/snapshot/AbstractAppDataServer.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.lib.appdata.snapshot;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+import com.datatorrent.common.experimental.AppData;
+import com.datatorrent.lib.appdata.StoreUtils;
+
+/**
+ * This is an operator that lays the framework of serving data to queries coming from an embeddable input operator.
+ * Subclasses are expected to implement the processQuery method for the logic of handling the query. Note that
+ * processQuery cannot directly emit to the operator's output port because it's called from the thread of the
```

Comment:
Comment wrt emit may be a bit hard to understand when referring to "input operator" because it isn't an operator/stream? Maybe refer to embedded query provider instead?